### PR TITLE
Add Utils DecodeWriteToken

### PR DESCRIPTION
**Pull request by elv-serban**
Example:

`tqw__HSa9UBm1dHmrLsz5GUu3D9NVscUshTBjW3TP2BckZzGzNHPtMEnWbpCDsiqB971qdYpy1oSemHMqTR1YsMsç`

Returns:

```
{
  writeTokenType: 'tqw__',
  tokId: '0xe45daaca900e5987d9de4ad769713e1a',
  qid: 'iq__4ZKDfwY1EtDbJvsJTDZLrTDb6H6b',
  nid: 'inod2vU5Lc28S9pcx96m2gbP4YMYeGRs'
}
```

Compared to `elv` command line:

```
elv tools decode tqw__HSa9UBm1dHmrLsz5GUu3D9NVscUshTBjW3TP2BckZzGzNHPtMEnWbpCDsiqB971qdYpy1oSemHMqTR1YsMs
==> tqw__HSa9UBm1dHmrLsz5GUu3D9NVscUshTBjW3TP2BckZzGzNHPtMEnWbpCDsiqB971qdYpy1oSemHMqTR1YsMs:
    type:   content write token
    bytes:  0xe45daaca900e5987d9de4ad769713e1a
    qid:    iq__4ZKDfwY1EtDbJvsJTDZLrTDb6H6b
    nid:    inod2vU5Lc28S9pcx96m2gbP4YMYeGRs
```

### CODE REVIEW

File: src/Utils.js [hunk 1]
```diff
@@ -150,6 +150,51 @@ const Utils = {
...
+
+    // Strip 5 character type
+    writeTokenType = writeToken.slice(0, 5);
+    writeToken = writeToken.slice(5);
+
+    // Decode base58 payload
+    let bytes = Utils.FromB58(writeToken);
+
+    // Parse QID
+    const qidBytes = bytes.slice(2, 22); // 1 byte size + 1 byte type
```
Comment (elv-marc) [inline src/Utils.js:177]: So the size here is assumed to always be one 1 byte and the qid to always be on 21. Is that a valid assumption?  I have similar code, that does not make those asumptions:  let ExtractQIDFromWriteToken = function(versionHash) {
  let Utils = Client.utils;
  versionHash = versionHash.replace(/^.*__/,"");

  // Decode base58 payload
  let bytes = Utils.FromB58(versionHash);


  // Determine size of varint content size
  let sizeLength = 0;
  while(bytes[sizeLength] >= 128) {
    sizeLength++;
  }
  sizeLength++;

  // Remove size
  const sizeBytes = bytes.slice(0, sizeLength);
  const size = VarInt.decode(sizeBytes);
  console.log({sizeBytes, size, sizeLength})
  bytes = bytes.slice(sizeLength, size+sizeLength);
  // Remaining bytes is object ID
  const objectId = "iq__" + Utils.B58(bytes.slice(1));

  // Part hash is B58 encoded version hash without the ID
  
  return    objectId

}

Comment (elv-gilles) [reply src/Utils.js:177]: The assumption is valid as long as IDs are ethereum addresses.

Here is another version that uses `VarInt.decode`. 
Note that the 3rd case in the unit-test fails with the code in the PR and not with the code below. 

```
  /**
   * Decode the specified write token into its component parts
   * Format:
   * - 1 + 1 + 20 bytes - content ID (size, type, value)
   * - 1 + 1 + 20 bytes - node ID (size, type, value)
   * - 1 + 16 bytes - token unique, random ID (size, value)
   *
   * @param writeToken
   *
   * @returns {Object} - Components of the write token.
   */
  DecodeWriteToken: (writeToken) => {
    if(!(writeToken.startsWith("tqw__"))) {
      throw new Error(`Invalid write token: "${writeToken}"`);
    }

    // Strip 5 character type
    writeTokenType = writeToken.slice(0, 5);
    writeToken = writeToken.slice(5);

    // Decode base58 payload
    let bytes = Utils.FromB58(writeToken);

    let qid;
    let nid;
    let tokId;

    function decode(prefix, index) {
      let bsize = VarInt.decode(bytes,0);
      let offset = VarInt.decode.bytes;
      let theBytes;

      switch(index) {
        case 0:
          theBytes = bytes.slice(offset+1, bsize+1); // 1 byte size + 1 byte type
          qid = "iq__" + Utils.B58(theBytes);
          break;
        case 1:
          theBytes = bytes.slice(offset+1, bsize+1); // 1 byte size + 1 byte type
          nid = "inod" + Utils.B58(theBytes);
          break;
        case 2:
          theBytes = bytes.slice(offset, bsize+1);
          tokId = "0x" + theBytes.toString("hex");
      }

      bytes = bytes.slice(bsize+1);
    }

    const ids = ["iq__", "inod", "0x"];
    ids.forEach(decode);

    return {
      writeTokenType,
      tokId,
      qid,
      nid
    };
  },
```

unit-test:

```
  test("Decode Write Token", () => {
    let writeToken = "tqw__HSa9UBm1dHmrLsz5GUu3D9NVscUshTBjW3TP2BckZzGzNHPtMEnWbpCDsiqB971qdYpy1oSemHMqTR1YsMs";
    let token = Utils.DecodeWriteToken(writeToken);
    expect(token.writeTokenType).toEqual("tqw__");
    expect(token.nid).toEqual("inod2vU5Lc28S9pcx96m2gbP4YMYeGRs");
    expect(token.qid).toEqual("iq__4ZKDfwY1EtDbJvsJTDZLrTDb6H6b");
    expect(token.tokId).toEqual("0xe45daaca900e5987d9de4ad769713e1a");

    writeToken = "tqw__8UmhDD9cZah58THfAYPf3Shj9hVzfwT51Cf4ZHKpayajzZRyMwCPiSpfS5yqRZfjkDjrtXuRmDa";
    token = Utils.DecodeWriteToken(writeToken);
    expect(token.writeTokenType).toEqual("tqw__");
    expect(token.nid).toEqual("inod3Sa5p3czRyYi8GnVGnh8gBDLaqJr");
    expect(token.qid).toEqual("iq__99d4kp14eSDEP7HWfjU4W6qmqDw");
    expect(token.tokId).toEqual("0x00010203040506070809");

    writeToken = "tqw__RRceZrF4GMkYHyXy9pD7wqvSLYxXbHxfNRPoQ8M1HuLsgRytg3GvimtfzdniU";
    token = Utils.DecodeWriteToken(writeToken);
    expect(token.writeTokenType).toEqual("tqw__");
    expect(token.nid).toEqual("inodBCzMsgCFoE629MAQih53ZP");
    expect(token.qid).toEqual("iq__XkHB7orSPYxxZYH8YQwA6F");
    expect(token.tokId).toEqual("0x66d17d8bd7e23869");
  });
```

**Review by elv-gilles — COMMENTED** [review]
see my reply to Marc (where there's a unit-test)

**Comment by lukseven** [comment]
Guys, 
please note that the `tqw__` prefix (note the 2 underscores!) is a backward-compatibility hack we had to provide when we added qid and nid to content write token. In fact, the new token format has the prefix `tq__`, whereas the old "random-bytes-only" format was `tqw_`. At the time, however, elv-client-js assumed that all write tokens started with `tqw_` and the new token `tq__` broke it. As a workaround, I temporarily changed the new token to `tqw__` with 2 underscores, the idea being that we fix elv-client-js and then revert back to the correct `tq__` prefix.

Maybe its time to just do that!

See 
* original token change: https://github.com/qluvio/content-fabric/pull/1552
* compatibility hack: https://github.com/qluvio/content-fabric/pull/1574

**Comment by elv-serban** [comment]
Yes let's clean up this part.

We can make this function parse all types.  Gilles can you help get the
parsing done based on 'prefix' (type).

Once we push this version we can then remove the backward compatible hack

- Serban Simu
  President & Co-founder Eluvio, Inc.  -  http://eluv.io

**Pull request merged**